src/signalduino.h

```diff
@@ -96,9 +96,26 @@ void setup() {
     ; // wait for serial port to connect. Needed for native USB
   }
 
-  for (uint8_t i=4;i<10;i++) { // pullup on for unused pins
-    pinAsInputPullUp(i);
-  }
+  // defined states - pullup on for unused pins
+  #ifndef ARDUINO_RADINOCC1101
+    #ifdef CMP_CC1101
+      for (uint8_t i=4;i<9;i++) {
+        pinAsInputPullUp(i);
+      }
+      for (uint8_t i=14;i<20;i++) {
+        pinAsInputPullUp(i);
+      }
+    #else
+      for (uint8_t i=4;i<20;i++) {
+        if (i != 9) pinAsInputPullUp(i); // not LED
+      }
+    #endif
```

Comment:
Hier ist uns noch ein Fehler unterlaufen, der Arduino Nano ohne CC1101 verwendet andere Pins.
So funktioniert es wieder:

```
    #else // Arduino Nano without CC1101
      for (uint8_t i=3;i<20;i++) { // not PIN_RECEIVE
        if ( i != PIN_SEND && i != PIN_LED ) {pinAsInputPullUp(i);}
      }
    #endif
```

Reply:
Die Sache mit Pin_Led leuchtet mir ein, da wir den nie initialisieren.

Pin_Send setzen wir später auf einen definierten Stand.
Ist dir bewusst, wieso das zum Problem führt?

Reply:
> Die Sache mit Pin_Led leuchtet mir ein, da wir den nie initialisieren.

Mhmm, `pinAsOutput(PIN_LED);` erfolgt gleich nach dem die Pullups der nicht angeschlossenen Pins eingeschaltet werden.

> Pin_Send setzen wir später auf einen definierten Stand.
> Ist dir bewusst, wieso das zum Problem führt?

Ich sehe in der output.h nicht so ganz durch, wie die Befehle umgesetzt werden.
Mir ist der Fehler aufgefallen, weil der SIGNALduino mit der alten Variante dauernd sendete.
Ich habe LEDs an PIN_SEND und PIN_RECEIVE. Die LED an PIN_SEND leuchtete dann dauernd.
`pinAsOutput(PIN_SEND);` scheint offensichtlich nicht mehr das zu tun, was es soll, wenn vorher der Pullup eingeschaltet wurde.

Ich habe auch nicht verstanden, warum ich `{pinAsInputPullUp(i);}` unbedingt in geschweifte Klammern setzen musste. Ohne funktionierte es nicht.

Reply:
Also ich habe mir das auch nochmal angesehen - mit dem Scope kann ich nicht sehen, dass die internen Pullups irgendeinen Einfluss auf das Signal haetten jedoch bringt es bei mir auch eine deutliche Verbesserung.

Ich habe ja als Fehlerursache die statische Varible in dem wait_Miso_rf() define in cc1101.h und entsprechend gefixt (den Pullrequest passe ich noch final an).

Um mal ein Gefuehl dafuer zu bekommen, habe ich den Code noch mit etwas mehr Debug versehen:

cc1101.c

```
void cc1101::sendFIFO(char *startpos, char *endpos) {
    cc1101_Select();                                // select CC1101
    sendSPI(CC1101_TXFIFO | CC1101_WRITE_BURST);    // send register address

    uint8_t val;
    for (char *i = startpos; i < endpos; i+=2) {
      val = hex2int(i[0]) * 16;
      val+= hex2int(i[1]);
      /* DBG_PRINTLN(val); // only to debug */
      sendSPI(val);    // send value
    }

    cc1101_Deselect();    //Wait for sending to finish (CC1101 will go to RX state automatically

    for(uint8_t i=0; i< 200;++i) {
      if( readReg(CC1101_MARCSTATE_REV00, CC1101_STATUS) != MarcStateTx)
        break;            //neither in RX nor TX, probably some error
      delay(1);
    }

    DBG_PRINTLN("waitMisoCnt");
    DBG_PRINTLN(waitMisoCnt);
    DBG_PRINTLN("waitMisoRfCnt");
    DBG_PRINTLN(waitMisoRfCnt);
}
```

In cc1101.h entsprechend das Miso wait gefixt (damit es stabil laeuft, aber mal waitMisoCnt und waitMisoRfCnt fuer Statistiken rein)

```
	#define wait_Miso()       { uint8_t miso_count = 255; while(isHigh(misoPin)) { waitMisoCnt++;   delay(1); if(miso_count == 0) return      ; miso_count--; } }    // wait until SPI MISO line goes low
	#define wait_Miso_rf()    { uint8_t miso_count = 255; while(isHigh(misoPin)) { waitMisoRfCnt++; delay(1); if(miso_count == 0) return false; miso_count--; } }    // wait until SPI MISO line goes low
```


Pullups abgeschaltet (also Dein Fix), 4 mal was gesendet:

```
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
6
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
8
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
10
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
12
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
```

Was man nun hier sieht: Bei jedem Senden wurde waitMisoRfCnt stetig um 2 erhoeht, d.h. es musste 2 mal auf MISO LOW gewartet werden, was vollkommen in Ordnung ist.


Jetzt machen wir das selbe Spiel wieder mit Pullups eingeschaltet:

```
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
23
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
40
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
57
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
send cmd detected 2
Adding xFSK message
rearrange beginptr
SN;locating data start:E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000; end:;
msg 1/1 part 0/0 repeats 1
waitMisoCnt
0
waitMisoRfCnt
74
SN;D=E3577F24FCBCEDA9F867EFFFFF1CA880DB034312560798100000;
```


Nun erhoeht es sich schon um 17 bei jedem send - also stoeren die Pullups vermutlich auf die Abfallzeit des Pegels von MISO.

Man sieht hier also deutlich, dass die Pullups abzuschalten richtig ist - dennoch: sind die eingeschaltet eigentlich nicht das Hauptproblem, dauert halt etwas laenger. 

Ursaechlich fuer das Problem ist nach wie vor die statische Variable in den miso waits. Man muss bedenken, wenn 255 auf 0 ist, wird immer abgebrochen.
Also mit Pullups
255, 253, 251,...,203,201,0 kaputt

Ohne Pullups
255, 238, 221, ... 51, 34, 0 kaputt
Es dauert hier einfach nur laenger.

Initial wusste ich ja nicht, was los ist und schrieb im Titel, dass ich beobachtet habe, dass es bei mir nur mal 15 klappt etwa... 255/17 = na? genau, 15 :)